### PR TITLE
Speedreader change icon behavior

### PR DESCRIPTION
**Pull request by keur**
Resolves https://github.com/brave/brave-browser/issues/16881, https://github.com/brave/brave-browser/issues/9262

## Submitter Checklist:

- [ ] I confirm that no security/privacy review [is needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or that I have [requested](https://github.com/brave/security/issues/new/choose) one
- [ ] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [ ] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [ ] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [ ] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [ ] Checked the PR locally: `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests`, `npm run lint`, `npm run gn_check`, `npm run tslint`
- [ ] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

### CODE REVIEW

File: browser/ui/views/speedreader/speedreader_icon_view.cc [hunk 1]
```diff
@@ -92,32 +76,44 @@ void SpeedreaderIconView::UpdateImpl() {
...
-                                               : IDS_DISTILL_PAGE);
+  int id;
+  const DistillState state = GetDistillState();
+  switch (state) {
+    case DistillState::kSpeedreaderMode:
+    case DistillState::kSpeedreaderOnDisabledPage:
+      id = IDS_SPEEDREADER_ICON_SPEEDREADER_SETTINGS;
+      break;
+    case DistillState::kReaderMode:
+      id = IDS_SPEEDREADER_ICON_TURN_OFF_READER_MODE;
```
Comment (keur) [inline browser/ui/views/speedreader/speedreader_icon_view.cc:111]: note that this state is assuming the behavior of https://github.com/brave/brave-core/pull/9393 being merged

File: browser/ui/views/speedreader/speedreader_icon_view.h [hunk 2]
```diff
@@ -6,19 +6,20 @@
 #ifndef BRAVE_BROWSER_UI_VIEWS_SPEEDREADER_SPEEDREADER_ICON_VIEW_H_
 #define BRAVE_BROWSER_UI_VIEWS_SPEEDREADER_SPEEDREADER_ICON_VIEW_H_
 
+#include "brave/browser/speedreader/speedreader_tab_helper.h"
 #include "chrome/browser/ui/views/page_action/page_action_icon_view.h"
 #include "content/public/browser/web_contents_observer.h"
 #include "ui/base/metadata/metadata_header_macros.h"
 
+using DistillState = speedreader::SpeedreaderTabHelper::DistillState;
```
Comment (iefremov) [inline browser/ui/views/speedreader/speedreader_icon_view.h:14]: pls move to .cc

**Review by iefremov — APPROVED** [review]

File: browser/ui/views/speedreader/speedreader_icon_view.h [hunk 3]
```diff
@@ -6,19 +6,20 @@
 #ifndef BRAVE_BROWSER_UI_VIEWS_SPEEDREADER_SPEEDREADER_ICON_VIEW_H_
 #define BRAVE_BROWSER_UI_VIEWS_SPEEDREADER_SPEEDREADER_ICON_VIEW_H_
 
+#include "brave/browser/speedreader/speedreader_tab_helper.h"
 #include "chrome/browser/ui/views/page_action/page_action_icon_view.h"
 #include "content/public/browser/web_contents_observer.h"
```
Comment (simonhong) [inline browser/ui/views/speedreader/speedreader_icon_view.h:11]: This can be deleted.

File: browser/ui/views/speedreader/speedreader_icon_view.h [hunk 4]
```diff
@@ -6,19 +6,20 @@
...
 #include "ui/base/metadata/metadata_header_macros.h"
 
+using DistillState = speedreader::SpeedreaderTabHelper::DistillState;
+
 namespace content {
 class NavigationHandle;
```
Comment (simonhong) [inline browser/ui/views/speedreader/speedreader_icon_view.h:17]: this is also not used?

**Review by simonhong — APPROVED** [review]

File: browser/ui/views/speedreader/speedreader_icon_view.h (hunk 2 above)
Comment (keur) [reply browser/ui/views/speedreader/speedreader_icon_view.h]: fixed

File: browser/ui/views/speedreader/speedreader_icon_view.h (hunk 3 above)
Comment (keur) [reply browser/ui/views/speedreader/speedreader_icon_view.h]: fixed.

File: browser/ui/views/speedreader/speedreader_icon_view.h (hunk 4 above)
Comment (keur) [reply browser/ui/views/speedreader/speedreader_icon_view.h]: fixed

File: browser/ui/views/speedreader/speedreader_icon_view.cc (hunk 1 above)
Comment (keur) [reply browser/ui/views/speedreader/speedreader_icon_view.cc:113]: merged 9393 and rebased before merge. resolving.

**Comment by keur** [comment]
Test failures fixed in https://github.com/brave/brave-core/pull/9400. Unrelated

**Pull request merged**